### PR TITLE
feat(table): sorting in data-table as web component 

### PR DESCRIPTION
**Pull request by theJohnnyMe**
**Describe pull-request**  
Make data table with sorting functionality

**Solving issue**  
_In case of GitHub issue add # plus the number of the issue (for example #123) OR if it is Azure then AB# and number of ticket_
Fixes: [AB#1787](https://dev.azure.com/scaniadigitaldesignsystem/3d9eb9c6-0045-473d-bc1d-e842dd779200/_workitems/edit/1787), [AB#1696](https://dev.azure.com/scaniadigitaldesignsystem/3d9eb9c6-0045-473d-bc1d-e842dd779200/_workitems/edit/1696), [AB#1780](https://dev.azure.com/scaniadigitaldesignsystem/3d9eb9c6-0045-473d-bc1d-e842dd779200/_workitems/edit/1780)

**How to test**  
_Add a description of how to test if possible_
1. Go to https://task-1696.d1g8v8mrkx992n.amplifyapp.com/?path=/story/component-table--web-component
2. Check Web Component under Table

**Additional context**  
Figma prototype: https://www.figma.com/proto/ZUT3Te7JVWggoukL9QGsPp/Components?page-id=4%3A6&node-id=948%3A99358&viewport=241%2C48%2C0.53&scaling=min-zoom&starting-point-node-id=948%3A99358&show-proto-sidebar=1

Figma design: https://www.figma.com/file/ZUT3Te7JVWggoukL9QGsPp/Components?node-id=4%3A6

### CODE REVIEW

**Comment by areblk** [comment]
So nice to see this as a web component! Nice animation on icon when one column is sorted and another column is sorted next!

I'd like to do some tests with dirtier data input (combined alphabetical and numerical values, like vehicle registration numbers, special characters and try to break strings acting as numericals). 

Here are notes on stuff I've tested so far, most things having to do with styling.

## Issues
1. Expected corner radius as illustrated here: https://dw1u92ia6bs84.cloudfront.net/Images/Image_1_c72095c012.png
Actual result: Incorrect corner radius on <th> when column is hovered or sorting is active. Only first top left and last top right corners should have radius of 4px. All other edges should have 0px radius.

corner-radius-1

corner-radius-2

corner-radius-3

2. Expected <th> to slightly overlap <caption> when paired as illustrated here: https://dw1u92ia6bs84.cloudfront.net/Images/Image_d339d5fdc4.png
Now, there is a very(!) small slightly lighter area below <caption> left bottom corner that looks a bit off (red arrow in image).
3. Expected whole column to change color on hovering <th> as illustrated here https://dw1u92ia6bs84.cloudfront.net/Images/Image_d339d5fdc4.png
Now, only <th> changes color on hover.
(Column hover isn't in the prototype though)
4. Expected hover on selected <th> to have background-color $grey-400 but it doesn't change
5. Expected data cells (<tr>) on-white to be $grey-50, and $grey-200 hover. Now, it has the same colors as on-grey ($white/$grey-100).

## Notes
6. Should sorting table be available as compact? Adding .sdds-table-compact currently doesn't affect <th>
7. Current table not scrollable as illustrated in docs (neither horizontally or vertically)
8. Current table not keyboard accessible
9. Should there be a way to reset table back to default without reloading the whole page?

File: components/src/components/table/table-component.stories.js [hunk 1]
```diff
@@ -0,0 +1,41 @@
...
+    label: {
+      name: 'Label',
+      type: 'string',
+      defaultValue: 'Label text',
+      description: 'Label text explains about dropdown',
+    },
+    bodyData: {
+      name: 'Body Data',
+      defaultValue:
+        '[{"eyeColor":"green","name":"Garrett","gender":"male","company":-3},{"eyeColor":"green","name":"Rodriguez","gender":"male","company":100},{"eyeColor":"green","name":"Minerva","gender":"female","company":0},{"eyeColor":"green","name":"Nona","gender":"female","company":235},{"eyeColor":"blue","name":"Barber","gender":"male","company":15}]',
```
Comment (areblk) [inline components/src/components/table/table-component.stories.js:20]: Where would this data present itself? Can't find it in storybook and Control = BodyData is empty

image

**Comment by theJohnnyMe** [comment]
@areblk 

_Should sorting table be available as compact? Adding .sdds-table-compact currently doesn't affect_
I have added bunch of addons for control of table now, compact design should be available now.

_Current table not scrollable as illustrated in docs (neither horizontally or vertically)_
Speaking with Alex this is something we would bring in as a feature in next iteration. This component is growing in complexity. Now we already have 3 web components in to bring all functionality.
 
_Current table not keyboard accessible_
I did small tweak, should be possible to move over head cells with tab and with enter do same as click.

_Should there be a way to reset table back to default without reloading the whole page?_
Nice idea, I think we could bring it up in next iteration. Maybe a button as part of the footer area or caption area. What do you think @alexlioliopoulos ?

Annika, I did not fix yet issues with round corners or even try to reproduce them. I will give it a look on Monday. Thanks for the great review. On the possibility to change test data in Storybook, unfortunately, it is not possible. JSON is complex data and it can not pass to the component in a classic way. I will spend time on Monday to try to see what can be done.

@areblk and @alexlioliopoulos - I have some ideas to release this one under pre-alpha or alpha tag. Why so? There is a still lot of functionality we want to implement here and the component is growing - 15 files changed in some way or some are newly created. With the release with the tag of alpha, people will have something to look and they will be informed that we are developing this one while being careful as we might bring in more changes. Let's talk on Monday about it.

File: components/src/components/table/table-component.stories.js (hunk 1 above)
Comment (theJohnnyMe) [reply components/src/components/table/table-component.stories.js]: This one won't be possible, unfortunately.

**Comment by areblk** [comment]
On the issue of whole column not changing color on hover. The design says that `sdds-body-cell` should change background-color on hovering it's `sdds-header-cell`. Now, that color change is implemented when header-cell is active, which is not in the design.

Will also look at new commits

**Comment by areblk** [comment]
New notes:

10. Design docs lists center and right text alignment as a Don't. If we're going to allow this, the docs on this topic needs to be updated. See 'Horizontal alignment' on https://digitaldesign.scania.com/components/data-tables/usage#do%E2%80%99s-and-don%E2%80%99ts
11. Storybook control `Vertical dividers = true` adds some pixels to the table with and pushes the content some pixels to the right. Possibly not an issue
12. How is column width value as percentage expected to work? How many columns can have it's width set as percentage? I can only have two values as percentage and get expected width. When setting three columns as 10% the first column is smaller than that. Reproduce: `no minimum with=true`, `column 1 width = 10%`, `column 2 width = 300px`, `column 3 width = 10%`, `column 4 width = 10%`
13. Even with no minimum width, a column shouldn't be allowed to be shorter than what's needed to fit the icons and it's margins, especially left margin.
image

**Comment by alexlioliopoulos** [comment]
> New notes:
> 
> 10. Design docs lists center and right text alignment as a Don't. If we're going to allow this, the docs on this topic needs to be updated. See 'Horizontal alignment' on https://digitaldesign.scania.com/components/data-tables/usage#do%E2%80%99s-and-don%E2%80%99ts
> 11. Storybook control `Vertical dividers = true` adds some pixels to the table with and pushes the content some pixels to the right. Possibly not an issue
> 12. How is column width value as percentage expected to work? How many columns can have it's width set as percentage? I can only have two values as percentage and get expected width. When setting three columns as 10% the first column is smaller than that. Reproduce: `no minimum with=true`, `column 1 width = 10%`, `column 2 width = 300px`, `column 3 width = 10%`, `column 4 width = 10%`
> 13. Even with no minimum width, a column shouldn't be allowed to be shorter than what's needed to fit the icons and it's margins, especially left margin.
>     image

10. Kinda true! So we have discussed about this and the decision is that left AND right alignment will be allowed in our data tables. Left alignment is the way to go for most cases, but for numerical columns, right alignment is the way to go (excel does the same). So this option needs to exist, but you are right @areblk , the docs need to be slightly updated. Could you add a task for that? Otherwise I ll do it. However, central alignment is not going to be an option, so @theJohnnyMe could you remove this?
11. You are right, but I think it does not affect the overall experience of the tables, right?
12. & 13. That's a very interesting one. I think that if possible, I'd like that if one header cell breaks to a new row when setting it to a small width, ALL other header cells get the same height, so that we get a consistent, harmonious feel.

**Comment by alexlioliopoulos** [comment]
9. I don't personally see why we might need the reset button. Perhaps let's leave it for up until someones comes with such a request?

**Comment by alexlioliopoulos** [comment]
Regarding pre-alpha, this is just for people to know that they should not be expecting a 100% stable state right? For me, yes let's do that.

**Comment by alexlioliopoulos** [comment]
5. on-white does not seem to be there, or am I missing something?

**Comment by alexlioliopoulos** [comment]
4. Correct, but no-biggie I'd say. For now at least. Priority is to have something that works out.

**Comment by alexlioliopoulos** [comment]
Overall - amazing job. Love what I am seeing. But also super good comments Annika, I would never be able to spot all of that myself.

**Comment by theJohnnyMe** [comment]
Thanks a lot to both of you for your detailed input, I will write here once I have some updates.

**Comment by theJohnnyMe** [comment]
What I have fixed until now:

- Table column can only accept left or right text align. If not specified, default is left.
- Entire column changes background color while we hover over its header cell.
- Round/sharp corners issues. 

I will keep on working and writing when I have new updates.

**Comment by areblk** [comment]
Input on labeling as pre-alpha or alpha. I think "alpha" is a better label for this component since it's more evolved than components currently labelled as pre-alpha (spinner, slider).

## :heavy_multiplication_x: Doesn't have to / will not be fixed 
- Hovering a selected <sdds-header-cell> should change background-color to $grey-400 (4) *"No biggie"*
- Table not scrollable in itself (7). *Next iteration*
- Reset-button (9). *No apparent use case*
- Vertical divider adds pixels to table (11). *Not enough impact*

## :heavy_check_mark: Fixed
- Column table-cells changes color when hovering header-cell (3)
- Possible to sort select and sort columns with keyboard `{tab}` and `{enter}`. Nice! (8)
- Only left and right text alignment is allowed, but not centered (~10).

## :construction: Party fixed
- Compact table reduces table-cell height, but doesn't affect header-cell height (6)
- Update documentation to reflect that left *and* right text alignment is allowed. AB#1195

## :warning: Issues not fixed
- Corner radius on Table Header (`<sdds-header-cell>`) when hovered or active (1) with slight overlap on <caption> (2)
- on-white theme variant, table-cell color default and on hover (5)
- Column allowed to be narrower than what is needed to fit header title and icon with margins (13)

## :new: New notes
15. I am a bit confused by the logic of width settings. For now, when `No minimum width = true` table max-width is 100vw. Explicit column width values are overwritten when viewport is to narrow. This means that user cannot set explicit column width that applies to all viewports. Is this what we want? I'm guessing that there is logical structure behind this that is not apparent to me.

16. If header cell contains title that is two rows or more, other header cells will have the same height, i.e. the button height will take up full height.

18. If table cell text alignment is right, should this also affect table-header? If we want to look like Excel, table-header is left aligned although numbers in column are right aligned. Right now, I think it looks kind of weird when table header is also right aligned because of the gap between the table headers. If we're using Excel as oracle, only table cells with numbers are right aligned, but header is left aligned. What do you say @alexlioliopoulos
image

Although, it looks a bit weird also when a wide column table-header is left aligned and table-cells are right aligned (example)
image

When combining with vertical dividers, it looks more like Excel
image


Excel
image

**Comment by alexlioliopoulos** [comment]
@areblk Regarding your point about left alignment for header title but right alignment for numerical cells, that's a good point, making the analogy to Excel. As you said yourself though, it looks quite odd when the width of the column is large, right? To be honest I think both ways could work, but this sounds like something that I might need to bring up with designers.

**Comment by alexlioliopoulos** [comment]
@areblk I particularly enjoyed the 
image

image

**Comment by theJohnnyMe** [comment]
@areblk and @alexlioliopoulos  - I fixed what I could for now. Please check it when you have time.
If some things are left, let's leave them for next week when I will work on multi select feature of the table.

The table now has:
- icon for sorting changes position option based on text alignment
- body cells change color based if prop white-background is true/false
- compact design is improved
- dividers are improved
- adding extra text to the header seems to be OK (goes in lines and makes the header element higher than usual)
- round corners should be fully fixed

Some of the comments on design are valid, but to be honest people have to read our docs and guidelines. Sure people can align one column right and make it 500px wide - will it look bad, yes, it that our fault - no. Everyone likes to have a good-looking website, well, please do read docs and do what we recommend. Use vertical dividers, lower the width - viola! :D 
I mean, we build trucks and educate future owners on how to use them properly. We can not stop them from misuse or faulty handling, something is on their side too.

**Comment by areblk** [comment]
If Alex is OK, I'm OK to send this to release. Status update below.

## :heavy_multiplication_x: Doesn't have to / will not be fixed 
- Hovering a selected <sdds-header-cell> should change background-color to $grey-400 (4) *"No biggie"*
- Table not scrollable in itself (7). *Next iteration*
- Reset-button (9). *No apparent use case*
- Vertical divider adds pixels to table (11). *Not enough impact*
- Decide if table header text alignment always left or following text alignment of table cells (18) *Alex will check with designers, edits will be in future iteration*

## :heavy_check_mark: Fixed
- Column table-cells changes color when hovering header-cell (3)
- Possible to sort select and sort columns with keyboard `{tab}` and `{enter}`. Nice! (8)
- Only left and right text alignment is allowed, but not centered (~10).
- Compact table reduces table-cell height, but doesn't affect header-cell height (6)
- Column allowed to be narrower than what is needed to fit header title and icon with margins (13)
- If header cell contains title that is two rows or more, other header cells will have the same height, i.e. the button height will take up full height (16)

## :construction: Party fixed
- Update documentation to reflect that left *and* right text alignment is allowed. [AB#1195](https://dev.azure.com/scaniadigitaldesignsystem/3d9eb9c6-0045-473d-bc1d-e842dd779200/_workitems/edit/1195)
- Document how it works to set explicit width values when using the nominwidth-prop (15)

## :warning: Issues not fixed
- Corner radius on Table Header (`<sdds-header-cell>`) when hovered or active (1) with slight overlap on <caption> (2). **Button corner radius still not perfect, but OK enough**
- on-white theme variant, table-cell color default and on hover (5)
- Sort under alpha/pre-alpha

**Review by areblk — APPROVED** [review]

**Review by alexlioliopoulos — APPROVED** [review]

**Comment by areblk** [comment]
> 
> ## ⚠️ Issues not fixed
>[...]
> * on-white theme variant, table-cell color default and on hover (5)
>[...]

I missed the control 'On white background' (which works as expected). Expected the table to change colour scheme when using storybook theme add-ons to set component class to "sdds-on-white-bg", like it works with Accordion and Cards.
image

**Pull request merged**